### PR TITLE
updatePRDescriptionsForCompleteTreeContainingCommit: Complete Stack Retreival

### PR DESCRIPTION
**Pull request by manyaagarwal**
**Summary**
Complete the PR stack for updating the PRs 

**Changelog**
- Add a parameter to the sttack.config.json i.e 'longLivedBranches: []` This currently stores the ref names for the long lived branches. A future update could be to store branch names and convert the branch name to ref name for getting the merge base commit. 
- Add a function to get the merge base commits in `GitSourceControl` 
- Updating the code to start from the first child of the merge base commit to update the whole PR Description Stack

### CODE REVIEW

File: src/stacker/ConcreteStacker.ts [hunk 1]
```diff
@@ -140,9 +142,43 @@ export class ConcreteStacker implements Stacker {
...
+            ...parentCommit.parentCommits,
+          ];
+          currentCommitHash = parentCommit.hash;
+        }
+      }
+    }
+    const nextCommits = [baseCommit];
     while (nextCommits.length) {
       // Push commit onto stack
+      console.log("NEXT COMMIT:", nextCommits);
```
Comment (saphal1998) [inline src/stacker/ConcreteStacker.ts]: Do we need to have this console log?

File: src/stacker/ConcreteStacker.ts [hunk 2]
```diff
@@ -160,19 +196,20 @@ export class ConcreteStacker implements Stacker {
       );
       nextCommits.push(...childCommits);
     }
+    console.log("Stack: ", stack);
```
Comment (saphal1998) [inline src/stacker/ConcreteStacker.ts]: Same goes for here, do we need this console log?

File: src/stacker/ConcreteStacker.ts [hunk 3]
```diff
@@ -140,9 +142,43 @@ export class ConcreteStacker implements Stacker {
...
+    while (parentCommitHashes.length) {
+      const parentCommitHash = parentCommitHashes.pop();
+      const baseCommitHash = mergeBaseCommitHashes.find(
+        (commitHash) => commitHash === parentCommitHash,
+      );
+      if (baseCommitHash) {
+        baseCommit = await this.sourceControl.getCommitByHash(
+          currentCommitHash,
+        );
+        console.log("CURRENT COMMIT:", baseCommit);
```
Comment (saphal1998) [inline src/stacker/ConcreteStacker.ts]: Is the console log here required?

**Review by saphal1998 — APPROVED** [review]
Looks good to me, I just had some general comments about the console logs

File: src/stacker/ConcreteStacker.ts (hunk 3 above)
Comment (manyaagarwal) [reply src/stacker/ConcreteStacker.ts]: Thanks for pointing these out! Removing them.

File: src/source-control/GitSourceControl.ts [hunk 4]
```diff
@@ -198,6 +198,33 @@ export class GitSourceControl implements SourceControl {
     this.repo = ourRepository;
   }
 
+  private getLLBFromConfig(): BranchName[] {
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]: Nit: this acronym is a little confusing hahah

```suggestion
  private getLongLivedBranchesFromConfig(): BranchName[] {
```

File: src/source-control/GitSourceControl.ts [hunk 5]
```diff
@@ -198,6 +198,33 @@ export class GitSourceControl implements SourceControl {
...
+    const configFileContents = fs
+      .readFileSync(`${this.repoPath}/sttack.config.json`)
+      .toString();
+    const { longLivedBranches } = JSON.parse(configFileContents);
+    return longLivedBranches;
+  }
+
+  async getMergeCommitByCommitHash(commit: Commit): Promise<CommitHash[]> {
+    const repo = await nodegit.Repository.open(this.repoPath);
+    const longlivedbranches: BranchName[] = this.getLLBFromConfig(); //TODO: Get long lived branches from sttack.config.json
```
Comment (taneliang) [inline src/source-control/GitSourceControl.ts]:
```suggestion
    const longlivedbranches: BranchName[] = this.getLongLivedBranchesFromConfig();
```

Probably need to run Prettier on this

File: src/source-control/SourceControl.ts [hunk 6]
```diff
@@ -39,6 +39,11 @@ export interface SourceControl {
    */
   getCommitByHash(hash: CommitHash): Promise<Commit | null>;
 
+  /**
+   * Get all the merge-base commit hashes on the long lived branches for a given commit 
+   */
+  getMergeCommitByCommitHash(commit: Commit) : Promise<CommitHash[]>;
```
Comment (taneliang) [inline src/source-control/SourceControl.ts]: "Merge commit" typically means something else to most people I think. Maybe this will be more accurate?

```suggestion
  getMergeBasesWithAllLongLivedBranches(commit: Commit) : Promise<CommitHash[]>;
```

File: src/stacker/ConcreteStacker.ts [hunk 7]
```diff
@@ -140,7 +142,38 @@ export class ConcreteStacker implements Stacker {
     commit: Commit,
   ): Promise<void> {
     const stack = [];
-    const nextCommits = [commit];
+    //Implement a complete version of the stack that start from the merge-base commit and also takes into consideration landed PRs
```
Comment (taneliang) [inline src/stacker/ConcreteStacker.ts]: Is this a TODO?

File: src/stacker/ConcreteStacker.ts [hunk 8]
```diff
@@ -43,8 +43,10 @@ export class ConcreteStacker implements Stacker {
     );
   }
 
-  loadRepositoryInformation(): void {
-    this.sourceControl.loadRepositoryInformation();
+  async loadRepositoryInformation(): Promise<void> {
+    await this.sourceControl.loadRepositoryInformation();
+    // TODO: Find PR info to load Pr info
+    // TODDO: SLoad PR info
```
Comment (taneliang) [inline src/stacker/ConcreteStacker.ts]: Oops, I typoed something

**Review by taneliang — COMMENTED** [review]
Looks great! Let me look through the code in more detail later and try to clarify some of the logic (the functions are long and I find it hard to follow on GitHub 😆)

**Review by taneliang — APPROVED** [review]
@manyaagarwal nice! Manually tested and it works 😍 

I've made some changes to the code organization, would appreciate if you could give it a once over before merging: [`4074f0a` (#41)](https://github.com/taneliang/stack-attack/pull/41/commits/4074f0abc46a610f1dbe070b938c9d898af81f63)

There's also this bug that may be encountered if `updatePRDescriptionsForCompleteTreeContainingCommit` encounters a merge commit. I've outlined it here: https://github.com/taneliang/stack-attack/pull/41/files#diff-7e24bfed562ffa8f53847965627f9162R177-R206. We can fix it in a future PR
